internal/server/v2/observation/calculation.go

```diff
@@ -17,56 +17,137 @@ package observation
 
 import (
 	"context"
+	"net/http"
 
+	pb "github.com/datacommonsorg/mixer/internal/proto"
 	pbv2 "github.com/datacommonsorg/mixer/internal/proto/v2"
 	"github.com/datacommonsorg/mixer/internal/server/cache"
+	"github.com/datacommonsorg/mixer/internal/server/resource"
 	"github.com/datacommonsorg/mixer/internal/store"
 )
 
-// CalculateObservationResponses returns a list of ObservationResponses based
-// on StatisticalCalculation formulas for the missing data in an input
-// ObservationResponse.
-func CalculateObservationResponses(
+// Computes a calculation for a variable and entity, based on a formula and input data.
+func Calculate(
 	ctx context.Context,
 	store *store.Store,
-	inputResp *pbv2.ObservationResponse,
 	cachedata *cache.Cache,
-) []*pbv2.ObservationResponse {
-	calculatedResponses := []*pbv2.ObservationResponse{}
-	for variable, variableObservation := range inputResp.ByVariable {
+	metadata *resource.Metadata,
+	httpClient *http.Client,
+	formula string,
+	variable string,
+	entity *pbv2.DcidOrExpression,
+	inputReq *pbv2.ObservationRequest,
+) (*pbv2.ObservationResponse, error) {
+	variableFormula, err := NewVariableFormula(formula)
+	if err != nil {
+		return nil, err
+	}
+	newReq := &pbv2.ObservationRequest{
+		Variable: &pbv2.DcidOrExpression{Dcids: variableFormula.StatVars},
+		Entity:   entity,
+		Date:     inputReq.GetDate(),
+		Value:    inputReq.GetValue(),
+		Filter:   inputReq.GetFilter(),
+		Select:   inputReq.GetSelect(),
+	}
+	variableResp, err := ObservationInternal(ctx, store, cachedata, metadata, httpClient, newReq)
+	if err != nil {
+		return nil, err
+	}
+	computeLeafObs(variableResp, variableFormula)
+	newResp := &pbv2.ObservationResponse{
+		ByVariable: map[string]*pbv2.VariableObservation{
+			variable: {},
+		},
+		Facets: map[string]*pb.Facet{},
+	}
+	calculatedResp, err := evalExpr(variableFormula.Expr, variableFormula)
+	if err != nil {
+		return nil, err
+	}
+	for entity, entityData := range calculatedResp {
+		facetObs := []*pbv2.FacetObservation{}
+		for facetId, facetData := range entityData {
+			if _, ok := newResp.Facets[facetId]; !ok {
+				newResp.Facets[facetId] = variableResp.Facets[facetId]
+			}
+			facetObs = append(facetObs, &pbv2.FacetObservation{
+				FacetId:      facetId,
+				Observations: facetData,
+				EarliestDate: facetData[0].GetDate(),
+				LatestDate:   facetData[len(facetData)-1].GetDate(),
+				ObsCount:     int32(len(facetData)),
+			})
+		}
+		if len(newResp.ByVariable[variable].ByEntity) == 0 {
+			newResp.ByVariable[variable].ByEntity = map[string]*pbv2.EntityObservation{}
+		}
+		newResp.ByVariable[variable].ByEntity[entity] = &pbv2.EntityObservation{
+			OrderedFacets: facetObs,
+		}
+	}
+	return newResp, nil
+}
+
+// Detects holes in a V2ObservationResponse and attempts to fill them using calculations.
+func MaybeCalculateHoles(
+	ctx context.Context,
+	store *store.Store,
+	cachedata *cache.Cache,
+	metadata *resource.Metadata,
+	httpClient *http.Client,
```

Comment:
Now as every function passes in these long list of args, should consolidate them into one class .. for later.